File: src/qc.cpp
#include "dcpwizard/qc.h"
#include "dcpwizard/verify.h"

#include <chrono>
#include <filesystem>
#include <spdlog/spdlog.h>

namespace fs = std::filesystem;

namespace dcpwizard
{

QCReport run_qc(const fs::path& dcp_dir)
{
  spdlog::info("Running QC on: {}", dcp_dir.string());
  auto start = std::chrono::steady_clock::now();
  QCReport report;
  report.standard = "SMPTE Bv2.1";

  if (!fs::exists(dcp_dir))
  {
    QCResult r;
    r.level = QCLevel::Error;
    r.code = "DIR_NOT_FOUND";
    r.message = "DCP directory does not exist";
    r.location = dcp_dir.string();
    report.results.push_back(r);
    report.passed = false;
    return report;
  }

  // Run structural verification
  auto verify = verify_dcp(dcp_dir);

  for (const auto& err : verify.errors)
  {
    QCResult r;
    r.level = QCLevel::Error;
    r.code = "STRUCTURE";
    r.message = err;
    report.results.push_back(r);
  }
  for (const auto& warn : verify.warnings)
  {
    QCResult r;
    r.level = QCLevel::Warning;
    r.code = "STRUCTURE";
    r.message = warn;
    report.results.push_back(r);
  }

  // Check MXF file sizes are reasonable
  for (const auto& entry : fs::directory_iterator(dcp_dir))
  {
    if (entry.path().extension() == ".mxf")
    {
      auto size = fs::file_size(entry.path());
      if (size < 1024)
      {
        QCResult r;
        r.level = QCLevel::Warning;
        r.code = "MXF_SIZE";
        r.message = "MXF file suspiciously small (" + std::to_string(size) + " bytes)";
        r.location = entry.path().filename().string();
        report.results.push_back(r);
      }
    }
  }

  auto end = std::chrono::steady_clock::now();
  report.duration_ms = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
  report.passed = verify.passed;

  spdlog::info("QC complete: {} ({}ms, {} notes)",
               report.passed ? "PASS" : "FAIL",
               report.duration_ms, report.results.size());
  return report;
}

} // namespace dcpwizard

File: rust/crates/dcpwizard-core/src/qc.rs
use serde::{Deserialize, Serialize};
use std::path::Path;

/// QC check severity level.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
pub enum QcLevel {
    #[default]
    Info,
    Warning,
    Error,
}

/// Single QC check result.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct QcResult {
    pub level: QcLevel,
    pub code: String,
    pub message: String,
}

/// Full QC report for a DCP.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct QcReport {
    pub dcp_path: String,
    pub passed: bool,
    pub results: Vec<QcResult>,
    pub error_count: usize,
    pub warning_count: usize,
    pub info_count: usize,
}

/// Run quality control checks on a DCP by delegating to dcpdoctor-core verify.
pub fn run_qc(dcp_dir: &Path) -> QcReport {
    let mut report = QcReport {
        dcp_path: dcp_dir.to_string_lossy().into_owned(),
        ..Default::default()
    };

    if !dcp_dir.exists() {
        report.results.push(QcResult {
            level: QcLevel::Error,
            code: "DCP_NOT_FOUND".into(),
            message: format!("DCP directory not found: {}", dcp_dir.display()),
        });
        report.error_count = 1;
        return report;
    }

    let verify_result = crate::verify::verify_dcp(dcp_dir);

    for err in &verify_result.errors {
        report.results.push(QcResult {
            level: QcLevel::Error,
            code: "VERIFY_ERROR".into(),
            message: err.clone(),
        });
    }

    for warn in &verify_result.warnings {
        report.results.push(QcResult {
            level: QcLevel::Warning,
            code: "VERIFY_WARNING".into(),
            message: warn.clone(),
        });
    }

    for note in &verify_result.info {
        report.results.push(QcResult {
            level: QcLevel::Info,
            code: "VERIFY_INFO".into(),
            message: note.clone(),
        });
    }

    report.error_count = verify_result.errors.len();
    report.warning_count = verify_result.warnings.len();
    report.info_count = verify_result.info.len();
    report.passed = verify_result.valid;

    // Additional structural checks
    let assetmap_exists =
        dcp_dir.join("ASSETMAP").exists() || dcp_dir.join("ASSETMAP.xml").exists();
    if !assetmap_exists {
        report.results.push(QcResult {
            level: QcLevel::Error,
            code: "MISSING_ASSETMAP".into(),
            message: "ASSETMAP not found".into(),
        });
        report.error_count += 1;
        report.passed = false;
    }

    let volindex_exists =
        dcp_dir.join("VOLINDEX").exists() || dcp_dir.join("VOLINDEX.xml").exists();
    if !volindex_exists {
        report.results.push(QcResult {
            level: QcLevel::Warning,
            code: "MISSING_VOLINDEX".into(),
            message: "VOLINDEX not found".into(),
        });
        report.warning_count += 1;
    }

    report
}
